docs/Proxies.md

```diff
@@ -0,0 +1,146 @@
+
+Another use-case for proxy contracts lies in contract upgrades. Since proxy contracts store the state and outsource the logic to an implementation contract, upgradeability is achieved by simply changing the reference to the implementation contract within the proxy contract. This allows developers to add features, update logic, and fix bugs without touching the state or the contract address to interact with the application. 
+
+### Proxy contract
+
+The [Proxy contract](../openzeppelin/upgrades/Proxy.cairo) includes two core methods:  
+
+1. The `__default__` method is a fallback method that redirects a function call and associated calldata to the implementation contract. 
+
+2. The `__l1_default__` method is also a fallback method; however, it redirects the function call and associated calldata to a layer one contract. In order to invoke `__l1_default__`, the original function call must include the library function `send_message_to_l1`. See Cairo's [Interacting with L1 contracts](https://www.cairo-lang.org/docs/hello_starknet/l1l2.html) for more information.
+
+Deploying a proxy contract requires only that the implementation contract is first deployed. The proxy contract accepts the implementation contract's address in the constructor. When interacting with the contract, function calls should be sent by the user to the proxy. The proxy's fallback function redirects the function call to the implementation contract to execute.
+
+
+### Implementation contract
+
+The implementation contract, also known as the logic contract, receives the redirected function calls from the proxy contract. The implementation contract should follow the [Extensibility pattern](../docs/Extensibility.md#the-pattern) and import directly from the [Proxy library](../openzeppelin/upgrades/library.cairo).
+ 
+
+The implementation contract should:
+- import `Proxy_initializer` and `Proxy_set_implementation`
+- initialize the proxy immediately after contract deployment.
+
+If the implementation is upgradeable, it should:
+- include a method to upgrade the implementation (i.e. `upgrade`)
+- use access control to protect the contract's upgradeability.
+
+The implementation contract should NOT:
+- deploy with a traditional constructor—use an initializer method instead that invokes `Proxy_initializer` and the implementation contract's initializer
+
+> Note that the imported `Proxy_initializer` includes a check the ensures the initializer can only be called once; however, `Proxy_set_implementation` does not include this check. It's up to the developers to protect their implementation contract's upgradeability with access controls such as `Proxy_only_admin`. 
```

Comment:
Perhaps we should document the whole API, so we can link to `Proxy_only_admin`

Reply:
Excellent idea